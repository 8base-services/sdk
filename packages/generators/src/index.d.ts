import { TableSchema } from '@8base/utils';

export const REPO_BRANCH_NAME: string;

export type GeneratorsCommonConfig = {
  deep: number,
  withMeta?: boolean,
};

export type GeneratorsConfig = {
  includeColumns?: string[],
} & GeneratorsCommonConfig;

export type GeneratorsData = {
  tablesList: TableSchema,
  tableName: string,
  screenName?: string,
};

export type ScreenTable = {
  id?: string,
  tableName: string,
  screenName: string,
  routeUrl?: string,
  tableFields?: string[],
  formFields?: string[],
};


export function replaceInitialApp(fsObject: Object, constants: {
  endpoint: string,
  authClientId: string,
  authDomain: string,
  appName: string,
}): Object;

export function generateScreen(data: {
  tablesList: TableSchema[],
  screen: ScreenTable,
  rootFile?: string,
}, config: GeneratorsCommonConfig): Object;
