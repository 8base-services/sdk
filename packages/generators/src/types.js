// @flow

export type GeneratorsConfig = {
  deep: number,
  includeColumns?: string[],
};
