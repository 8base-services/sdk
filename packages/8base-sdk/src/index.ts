export * from '@8base/apollo-client';
export * from '@8base/apollo-links';
export * from '@8base/auth';
export * from '@8base/utils';
export * from '@8base/validate';
