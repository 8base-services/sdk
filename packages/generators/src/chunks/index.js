// @flow
import ejs from 'ejs';

// $FlowIgnore
import fieldsInputsList from './fieldsInputsList.js.ejs';
// $FlowIgnore
import fieldsQueries from './fieldsQueries.js.ejs';
// $FlowIgnore
import fileField from './fileField.js.ejs';
// $FlowIgnore
import fieldsList from './fieldsList.js.ejs';
// $FlowIgnore
import addressField from './addressField.js.ejs';
// $FlowIgnore
import phoneField from './phoneField.js.ejs';
// $FlowIgnore
import createEditComponents from './createEditComponents.js.ejs';

export const chunks = {
  fieldsInputsList: (data?: Object) => ejs.render(fieldsInputsList, data),
  fieldsQueries: (data?: Object) => ejs.render(fieldsQueries, data),
  fileField: (data?: Object) => ejs.render(fileField, data),
  fieldsList: (data?: Object) => ejs.render(fieldsList, data),
  addressField: (data?: Object) => ejs.render(addressField, data),
  phoneField: (data?: Object) => ejs.render(phoneField, data),
  createEditComponents: (data?: Object) => ejs.render(createEditComponents, data),
};
