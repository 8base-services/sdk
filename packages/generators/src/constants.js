// @flow

/** Name of thi branch in repo https://github.com/8base/react-app-starter */
export const REPO_BRANCH_NAME = 'v1.1.0';
