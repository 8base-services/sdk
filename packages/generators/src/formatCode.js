// @flow
import prettier from 'prettier';

export const formatCode = (code: string) => prettier.format(code, {
  printWidth: 120,
  singleQuote: true,
  trailingComma: 'es5',
  parser: 'babel',
});
