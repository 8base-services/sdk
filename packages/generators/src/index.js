// @flow

export { generateCreateForm } from './generateCreateForm/createForm';
export { generateEditForm } from './generateEditForm/editForm';
export { generateDeleteForm } from './generateDeleteForm/deleteForm';
export { generateTable } from './generateTable/table';
export { generateIndex } from './generateIndex';

