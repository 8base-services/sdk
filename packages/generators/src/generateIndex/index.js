// @flow

import ejs from 'ejs';
import pluralize from 'pluralize';
import changeCase from 'change-case';
import { formatCode } from '../formatCode';

// $FlowIgnore
import index from './index.js.ejs';

export const generateIndex = ({ tableName, screenName }: *) => {
  const entityName = pluralize.singular(tableName);

  const tableGenerated = ejs.render(index, {
    changeCase,
    pluralize,
    tableName,
    entityName,
    screenName: screenName || entityName,
  });

  return formatCode(tableGenerated);
};

