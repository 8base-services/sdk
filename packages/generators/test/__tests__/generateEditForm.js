// @flow

import { generateEditForm } from '../../src';
import { tablesSchema } from '../__fixtures__/tablesSchema';


it('should generate update form by the table name', () => {
  const generatedEditForm = generateEditForm(tablesSchema, 'Properties');

  expect(generatedEditForm).toMatchSnapshot();
});
