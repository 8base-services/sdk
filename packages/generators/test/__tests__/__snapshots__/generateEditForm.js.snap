// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should generate update form by the table name 1`] = `
"import React from 'react';
import { memoizeWith, identity } from 'ramda';
import gql from 'graphql-tag';
import { Query, graphql } from 'react-apollo';
import { COUNTRY_CODE_OPTIONS } from '@8base/utils';
import { Form as FormLogic, Field, FieldArray } from '@8base/forms';
import {
  Dialog,
  Grid,
  Button,
  Form,
  Row,
  Column,
  Icon,
  Text,
  SelectField,
  InputField,
  CheckboxField,
  DateInputField,
  Label,
  ModalContext,
} from '@8base/boost';
import { FileInput } from '@8base/file-input';

const PROPERTY_UPDATE_MUTATION = gql\`
  mutation PropertyUpdate($data: PropertyUpdateInput!) {
    propertyUpdate(data: $data) {
      id
    }
  }
\`;
const LISTINGS_LIST_QUERY = gql\`
  query ListingsListList {
    listingsList {
      items {
        id
        _description
      }
    }
  }
\`;

const getRelationOptions = (items = []) => items.map(item => ({ value: item.id, label: item._description }));

const FileInputField = ({ input, meta, maxFiles, label, ...rest }) => (
  <Form.Field label={label} input={input} meta={meta}>
    <FileInput onChange={input.onChange} value={input.value} maxFiles={maxFiles} public={rest.public}>
      {({ pick, value }) => (
        <Row stretch alignItems=\\"center\\">
          <Button type=\\"button\\" onClick={pick} stretch color=\\"neutral\\">
            Choose Files
          </Button>
          <Text size=\\"sm\\">
            {value ? (Array.isArray(value) ? \`\${value.length} files selected\` : value.filename) : 'No files selected'}
          </Text>
        </Row>
      )}
    </FileInput>
  </Form.Field>
);

const ListFields = ({ fields, label, component, defaultValue = '', ...rest }) => (
  <React.Fragment>
    {fields.length > 0 && <Label>{label}</Label>}
    <Column gap=\\"sm\\" alignItems=\\"stretch\\">
      {fields.map((name, index) => (
        <Row gap=\\"sm\\" alignItems=\\"center\\">
          <Field {...rest} key={name} name={name} component={component} />
          <Icon color=\\"GRAY1\\" name=\\"Delete\\" cursor=\\"pointer\\" onClick={() => fields.remove(index)} />
        </Row>
      ))}
      <Text color=\\"GRAY1\\" cursor=\\"pointer\\" onClick={() => fields.push(defaultValue)}>
        Add {label}
      </Text>
    </Column>
  </React.Fragment>
);

const AddressField = ({ label, name }) => (
  <Column gap=\\"none\\" alignItems=\\"stretch\\">
    {label && <Label kind=\\"secondary\\">{label}</Label>}
    <Grid.Layout
      columns=\\"100px 1fr 1fr\\"
      autoRows=\\"auto\\"
      gap=\\"sm\\"
      areas={[
        ['country', 'country', 'country'],
        ['street1', 'street1', 'street1'],
        ['street2', 'street2', 'street2'],
        ['zip', 'city', 'state'],
      ]}
    >
      <Grid.Box area=\\"country\\">
        <Field name={\`\${name}.country\`} component={InputField} placeholder=\\"Country\\" />
      </Grid.Box>
      <Grid.Box area=\\"street1\\">
        <Field name={\`\${name}.street1\`} component={InputField} placeholder=\\"Address Line 1\\" />
      </Grid.Box>
      <Grid.Box area=\\"street2\\">
        <Field name={\`\${name}.street2\`} component={InputField} placeholder=\\"Address Line 2\\" />
      </Grid.Box>
      <Grid.Box area=\\"zip\\">
        <Field name={\`\${name}.zip\`} component={InputField} placeholder=\\"Postal/Zip\\" />
      </Grid.Box>
      <Grid.Box area=\\"city\\">
        <Field name={\`\${name}.city\`} component={InputField} placeholder=\\"City\\" />
      </Grid.Box>
      <Grid.Box area=\\"state\\">
        <Field name={\`\${name}.state\`} component={InputField} placeholder=\\"State\\" />
      </Grid.Box>
    </Grid.Layout>
  </Column>
);

const PhoneField = ({ label, name }) => (
  <Column gap=\\"none\\" alignItems=\\"stretch\\">
    {label && <Label kind=\\"secondary\\">{label}</Label>}
    <Grid.Layout columns=\\"120px 1fr\\" autoRows=\\"auto\\" gap=\\"lg\\">
      <Grid.Box>
        <Field name={\`\${name}.code\`} component={SelectField} placeholder=\\"Code\\" options={COUNTRY_CODE_OPTIONS} />
      </Grid.Box>
      <Grid.Box justifyContent=\\"flex-end\\">
        <Field name={\`\${name}.number\`} component={InputField} placeholder=\\"Number\\" />
      </Grid.Box>
    </Grid.Layout>
  </Column>
);

const ehnhancer = graphql(PROPERTY_UPDATE_MUTATION, {
  name: 'propertyUpdate',
  options: {
    refetchQueries: ['PropertiesTableContent'],
    context: {
      TOAST_SUCCESS_MESSAGE: 'Property successfuly updated',
    },
  },
});

const PropertyEditDialog = ehnhancer(
  class PropertyEditDialog extends React.PureComponent {
    static contextType = ModalContext;

    updateOnSubmit = memoizeWith(identity, id => async data => {
      await this.props.propertyUpdate({ variables: { data: { ...data, id } } });

      this.context.closeModal('PROPERTY_EDIT_DIALOG_ID');
    });

    onClose = () => {
      this.context.closeModal('PROPERTY_EDIT_DIALOG_ID');
    };

    renderFormContent = ({ handleSubmit, invalid, submitting, pristine }) => (
      <form onSubmit={handleSubmit}>
        <Dialog.Header title=\\"Edit Property\\" onClose={this.onClose} />
        <Dialog.Body scrollable>
          <Grid.Layout gap=\\"sm\\" stretch>
            <Grid.Box>
              <AddressField name=\\"address\\" label=\\"Address\\" />
            </Grid.Box>
            <Grid.Box>
              <PhoneField name=\\"phone\\" label=\\"Phone\\" />
            </Grid.Box>
            <Grid.Box>
              <FieldArray name=\\"pictures\\">
                {({ fields }) => <ListFields fields={fields} label=\\"Pictures\\" component={FileInputField} />}
              </FieldArray>
            </Grid.Box>
            <Grid.Box>
              <Field name=\\"bedrooms\\" label=\\"Bedrooms\\" type=\\"number\\" component={InputField} />
            </Grid.Box>
            <Grid.Box>
              <Field name=\\"title\\" label=\\"title\\" component={InputField} />
            </Grid.Box>
            <Grid.Box>
              <Field name=\\"description\\" label=\\"Description\\" component={InputField} />
            </Grid.Box>
            <Grid.Box>
              <Field name=\\"sqFootage\\" label=\\"Sq Footage\\" type=\\"number\\" component={InputField} />
            </Grid.Box>
            <Grid.Box>
              <Field name=\\"bathrooms\\" label=\\"Bathrooms\\" type=\\"number\\" component={InputField} />
            </Grid.Box>
            <Grid.Box>
              <Field name=\\"garage\\" label=\\"Garage\\" component={CheckboxField} />
            </Grid.Box>
            <Grid.Box>
              <Field name=\\"pool\\" label=\\"Pool\\" component={CheckboxField} />
            </Grid.Box>
            <Grid.Box>
              <Query query={LISTINGS_LIST_QUERY}>
                {({ data, loading }) => (
                  <Field
                    name=\\"listings\\"
                    label=\\"Listings\\"
                    multiple={true}
                    component={SelectField}
                    placeholder=\\"Select a listing\\"
                    loading={loading}
                    options={loading ? [] : getRelationOptions(data.listingsList.items)}
                    stretch
                  />
                )}
              </Query>
            </Grid.Box>
          </Grid.Layout>
        </Dialog.Body>
        <Dialog.Footer>
          <Button color=\\"neutral\\" variant=\\"outlined\\" disabled={submitting} onClick={this.onClose}>
            Cancel
          </Button>
          <Button color=\\"primary\\" type=\\"submit\\" disabled={pristine || invalid} loading={submitting}>
            Update Property
          </Button>
        </Dialog.Footer>
      </form>
    );

    renderForm = ({ args }) => {
      return (
        <FormLogic
          type=\\"UPDATE\\"
          tableSchemaName=\\"Properties\\"
          onSubmit={this.updateOnSubmit(args.initialValues.id)}
          initialValues={args.initialValues}
        >
          {this.renderFormContent}
        </FormLogic>
      );
    };

    render() {
      return (
        <Dialog id={'PROPERTY_EDIT_DIALOG_ID'} size=\\"sm\\">
          {this.renderForm}
        </Dialog>
      );
    }
  }
);

export { PropertyEditDialog };
"
`;
